src/cluster_legacy.c

```diff
@@ -4394,7 +4399,7 @@ void clusterSendFailoverAuthIfNeeded(clusterNode *node, clusterMsg *request) {
 
     /* We can vote for this replica. */
     server.cluster->lastVoteEpoch = server.cluster->currentEpoch;
-    node->replicaof->voted_time = mstime();
+    if (!force_ack) node->replicaof->voted_time = mstime();
```

Comment:
I don't get this, don't we still want to fence off normal failovers if we are triggering a manual failover?

Reply:
Do you mean both normal failover and the manual failover happen in the same time? This must happen in a force or takeover path? I think it might be rare, adding this is because i am worried that after a manual failover, if the primary is down quickly, then the normal failover will not be able to finish. I want all the manual failover bypass this logic.

Reply:
Yes, if you send a manual failover it won't reset the voted_time, so an automatic failover might come in and the node will also send a vote for the newly requested node? The force_ack path is already ignored the voted_time, so why wouldn't we update the voted_time here to prevent a future automatic failover from running?

Reply:
ok. i think i got your point, i will remove this change, i will try to think of a test case to cover it.

Reply:
This is the problem I want to solve, this is why i add this force_ack check. It is odd that i lost the context and did not put it in the TCL tests. The test case will do two manual failover, and the third one is the automatic failover, in here, if we remove the force_ack check, the automatic failover won't get the votes, since other primary has vote for the old primary in the first manual failover. 
```
start_cluster 3 1 {tags {external:skip cluster} overrides {cluster-ping-interval 1000 cluster-node-timeout 5000}} {
    test "xxxx" {
        R 3 cluster failover
        wait_for_condition 1000 50 {
            [s 0 role] eq {slave} &&
            [s -3 role] eq {master}
        } else {
            fail "The first failover does not happen"
        }
        wait_for_cluster_propagation

        R 0 cluster failover
        wait_for_condition 1000 50 {
            [s 0 role] eq {master} &&
            [s -3 role] eq {slave}
        } else {
            fail "The second failover does not happen"
        }
        wait_for_cluster_propagation

        # Let R 3 trigger the automatic failover
        pause_process [srv 0 pid]
        wait_for_cluster_state fail
        after 2000
    }
} ;# start_cluster
```

> If you send a manual failover it won't reset the voted_time, so an automatic failover request from another replica might come in and then this primary will also send a vote for the new replica that has requested it? This might cause the manual failover to get reverted and cause churn in the system.

i try this one, it is hard to write the test case, if they are happen in the same time, it is likely they do the election in the same epoch, so only one can win. If the automatic failover is the after one, it use a old epoch and it cant win the election. Or the manual failover win the election and broadcast the new config, the automatic failover also dont have the chance to win the election. So this look safe for me.

Reply:
I actually don’t see why this wait is necessary for automatic failovers, let alone manual ones. Does anyone have concrete context on why [this commit](https://github.com/valkey-io/valkey/commit/a445aa30a0fb72bc54da083998b3458fad92820f) introduced it? It doesn’t seem tied to any real-world use case or operational challenge.

This enforced wait significantly slows down recovery in scenarios where multiple primaries fail close together—such as during an AZ outage or a major network partition. 

Note that we already debounce transient primary failures with `cluster-node-timeout`, ensuring failover is only triggered after sustained outages. Election timing is naturally staggered by replica spacing, making the likelihood of simultaneous elections from replicas of the same shard very low. The one-vote-per-epoch rule further throttles retries and ensures orderly elections. On top of that, quorum-based primary failure confirmation, cluster-state convergence, and slot ownership validation are all built into the process.

If the concern is about stability during failovers, these existing mechanisms already handle it effectively IMO. Having this global wait not only fails to provide additional protection but actively hurts recovery speed, particularly in high-availability setups where responsiveness is critical.

Please let me know if I miss some specific edge cases that this wait was meant to address. Otherwise, I think we should move forward with removing it for both automatic and manual failovers.

Reply:
I'd also prefer to remove it completely, if there is other specific edge cases, we should use other way to fix it instead of this limitation (the slow down thing is suck i think and has never been documented).

Reply:
> If the concern is about stability during failovers, these existing mechanisms already handle it effectively IMO. Having this global wait not only fails to provide additional protection but actively hurts recovery speed, particularly in high-availability setups where responsiveness is critical.

This isn't a global wait, it's only against the specific primary. It's to prevent double failovers.

My general understanding, is if a primary just took over we don't want someone else to try to take over and give the new primary some amount of time to take over. I have not seen this issue though, it might have been over optimizing? The double failure mode, where a node fails and then another node fails within the nodetimeout also doesn't seem that common either though.

I guess I agree with removing it. It seems like it will make the code a lot simpler.